identity_iota/src/did/document.rs

```diff
@@ -1,145 +1,193 @@
+use core::{
+    convert::TryFrom,
+    fmt::{Debug, Display, Error as FmtError, Formatter, Result as FmtResult},
+    ops::{Deref, DerefMut},
+};
 use identity_core::{
-    common::{AsJson as _, Object, SerdeInto as _, Value},
-    did::DIDDocument,
-    key::{KeyData, KeyRelation, KeyType, PublicKey},
-    utils::{decode_b58, encode_b58},
+    common::{OneOrMany, ToJson as _},
+    did::{DIDDocument as Document, DIDDocumentBuilder as DocumentBuilder, DID},
+    diff::Diff as _,
+    key::{KeyRelation, KeyType, PublicKey},
 };
-use identity_crypto::{Ed25519, SecretKey, Sign as _, Verify as _};
+use identity_crypto::{KeyPair, SecretKey};
+use identity_proof::{signature::jcsed25519signature2020, LdDocument, LdSignature, SignatureOptions};
 use serde::{Deserialize, Serialize};
 
 use crate::{
-    did::{DIDDiff, DIDProof, TangleDocument},
+    did::{DIDDiff, IotaDID, LdDiffRead, LdDiffWrite},
     error::{Error, Result},
 };
 
-impl TangleDocument for DIDDocument {
-    fn sign_diff_unchecked(&self, diff: &mut DIDDiff, secret: &SecretKey) -> Result<()> {
-        // Get the first authentication key from the document.
-        let key: &PublicKey = self
-            .resolve_key(0, KeyRelation::Authentication)
-            .ok_or(Error::InvalidAuthenticationKey)?;
+#[derive(Clone, PartialEq, Deserialize, Serialize)]
+#[serde(try_from = "Document", into = "Document")]
+pub struct IotaDocument {
+    document: Document,
+}
+
+impl IotaDocument {
+    pub fn generate_ed25519_keypair() -> KeyPair {
+        jcsed25519signature2020::new_keypair()
+    }
 
-        // Reset the proof object in the diff.
-        diff.proof = DIDProof::new(key.id().clone());
+    pub fn try_from_document(document: Document) -> Result<Self> {
+        // The DID document MUST have a well-formed IOTA DID
+        let did: IotaDID = IotaDID::try_from_did(document.did().clone())?;
 
-        // Create a signature from the diff JSON.
-        let signature: String = sign_canonicalized(diff, key.key_type(), secret)?;
+        // The DID document MUST have an authentication key that matches the DID
+        let key: Vec<u8> = LdDocument::resolve_key(&document, 0.into())?;
+        let tag: String = IotaDID::encode_key(&key);
 
-        // Update the diff proof with the encoded signature.
-        diff.proof.signature = signature;
+        // The DID tag MUST equal Base58( Blake2b-256( authentication-key ) )
+        if did.method_id() != tag {
+            return Err(Error::InvalidAuthenticationKey);
+        }
 
-        Ok(())
+        Ok(Self { document })
     }
 
-    fn verify_diff_unchecked(&self, diff: &DIDDiff) -> Result<()> {
-        // TODO: Validate diff.id
-        // TODO: Validate diff.prevMsg
+    pub fn new(did: IotaDID, authentication: PublicKey) -> Result<Self> {
```

Comment:
`authentication` should be validated and `id/controller` should line up with `did`